### PR TITLE
[FIX]: issue #171 favicon counter is buggy on first load

### LINKED ISSUE #171: [BUG] Favicon Counter on First Load
hi @avalla tagging you here as it's your feature and idk your discord. We saw an issue where harvestable crops upon loading the game are not counted. It works when you plant and wait for the timer but not for already harvestable crops.

image

### PR DESCRIPTION
**Pull request by avalla**
# Description

Refactored AppIconContext harvestable counter in favicon.

Fixes #171

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

- harvesting and seeding;
- alter INITIAL_FARM in constants.ts

### CODE REVIEW

**Comment by Pranav-yadav** [comment]
Yep @avalla , even I've seen this bug on initial load.

**Comment by avalla** [comment]
it should be fixed with this PR

btw it will not be better if there is no counter, just a symbol shown when counter is greater than zero? i think that is useful to know if there is something to harvest, not how many fields

**Comment by j923sneaks** [comment]
It works on my end.

I noticed however that `updateHarvestable()` is called 9 times in `Soil.tsx` (since there are 9 plants initially). That means you execute `.filter()` 9 times even without `total` changing.
image

Maybe we can still optimize this, though I don't have any solutions to offer at this moment...

**Comment by avalla** [comment]
@j923sneaks

now there is a debounce function, please could you retry using  commit b6e4469 ?

Thanks!!

**Comment by j923sneaks** [comment]
@avalla yep much better! tested with planting as well.

These were separated by zones
image

> btw it will not be better if there is no counter, just a symbol shown when counter is greater than zero? i think that is useful to know if there is something to harvest, not how many fields

I don't think there's much difference in the implementation. If just a symbol, you'll use `.some()` instead of `.filter().length` which still iterates the array.

**Comment by avalla** [comment]
@j923sneaks 

Yes it will be almost the same! I think that the counter gives too many useless information

It will be similar to this (icon is not good, I know :))

Schermata 2022-01-26 alle 11 25 25

**Review by adamhannigan — APPROVED** [review]
Nice work 👍 Just a small few suggestions.


@avalla Also, I've noticed that the counter still shows for crops that are blocked by the Goblins - this was not introduced by this PR but something we will need to consider in the future.

This might be a bit confusing to players as they have the favicon showing, but cannot actually harvest anything as the Goblin is preventing them

Screen Shot 2022-01-26 at 11 12 35 am

File: src/features/crops/AppIconProvider.tsx [hunk 1]
```diff
@@ -2,31 +2,54 @@
...
  */
 
-import React from "react";
-import { useState } from "react";
+import React, {useContext} from "react";
 import Tinycon from "tinycon";
+import {CROPS} from "features/game/types/crops";
+import {getTimeLeft} from "lib/utils/time";
+import {Context} from "features/game/GameProvider";
+import {useActor} from "@xstate/react";
```
[the remark this answers is not in the document]
Comment (avalla) [reply src/features/crops/AppIconProvider.tsx:10]: you're right, sorry :(

File: src/features/crops/AppIconProvider.tsx [hunk 2]
```diff
@@ -2,31 +2,54 @@
...
+export const AppIconProvider: React.FC = ({ children }) => {
+  const { gameService } = useContext(Context);
+  const [
+    {
+      context: { state },
+    },
+  ] = useActor(gameService);
+
+  const updateHarvestable = debounce(() : void => {
+    console.log(Date.now(), 'Update')
```
[the remark this answers is not in the document]
Comment (avalla) [reply src/features/crops/AppIconProvider.tsx:46]: right, twice

**Comment by avalla** [comment]
> Nice work 👍 Just a small few suggestions.
> 
> @avalla Also, I've noticed that the counter still shows for crops that are blocked by the Goblins - this was not introduced by this PR but something we will need to consider in the future.
> 
> This might be a bit confusing to players as they have the favicon showing, but cannot actually harvest anything as the Goblin is preventing them
> 
> Screen Shot 2022-01-26 at 11 12 35 am

@j923sneaks 

to detect if goblin covers the field we should check inventory and Field, right?

**Comment by j923sneaks** [comment]
> to detect if goblin covers the field we should check inventory and Field, right?

hi @avalla i think you can copy the logic inside `plant.ts` and `harvest.ts` for checking goblins 😄 

basically as long as you crafted the food, it unlocks the zone. What I don't know is if they're consumable or just craft and let it be in the inventory. Can you confirm, @adamhannigan ?

**Comment by adamhannigan** [comment]
@j923sneaks @avalla Yes, this is based on whether an item exists in the inventory. For instance, if you have pumpkin soup you have access to the second set of fields.

These items are non consumable/burnable, they just sit in your inventory

**Comment by adamhannigan** [comment]
@j923sneaks @avalla I will merge this in now. The locked fields is a bit out of scope for this issue. We can address it in another PR :)

**Pull request merged**